src/disk_cache.rs

```diff
@@ -0,0 +1,955 @@
+    object_store: Arc<dyn ObjectStore>,
+    part_size: usize, // expected to be aligned with mb or kb
+}
+
+impl DiskCachedObjectStore {
+    pub fn new(
+        object_store: Arc<dyn ObjectStore>,
+        root_folder: std::path::PathBuf,
+        part_size: usize,
+    ) -> Self {
+        assert!(part_size % 1024 == 0);
+
+        Self {
+            object_store,
+            root_folder,
+            part_size,
+        }
+    }
+
+    async fn cached_head(&self, location: &Path) -> object_store::Result<ObjectMeta> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+        match entry.read_meta().await {
+            Ok(Some(meta)) => Ok(meta),
+            _ => {
+                let meta = self.object_store.head(location).await?;
+                entry.save_meta(&meta).await.ok();
+                Ok(meta)
+            }
+        }
+    }
+
+    async fn cached_get_opts(
+        &self,
+        location: &Path,
+        opts: GetOptions,
+    ) -> object_store::Result<GetResult> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+
+        let meta = self.maybe_prefetch_range(&entry, &opts.range).await?;
+        let range = self.canonicalize_range(opts.range, meta.size)?;
+        let parts = self.split_range_into_parts(range.clone());
+
+        // read parts, and concatenate them into a single stream. please note that some of these part may not be cached,
+        // we'll still fallback to the object store to get the missing parts.
+        let futures = parts
+            .into_iter()
+            .map(|(part_id, range_in_part)| self.read_part(location, part_id, range_in_part))
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes: Default::default(),
+            payload: GetResultPayload::Stream(result_stream),
+        })
+    }
+
+    // if an object is not cached before, maybe_prefetch_range will try to prefetch the object from the
+    // object store and save the parts into the local disk cache. the prefetching is helpful to reduce the
+    // number of GET requests to the object store, it'll try to aggregate the parts among the range into a
+    // single GET request, and save the related parts into local disks together.
+    // when it sends GET requests to the object store, the range is expected to be ALIGNED with the part
+    // size.
+    async fn maybe_prefetch_range(
+        &self,
+        entry: &DiskCacheEntry,
+        range: &Option<GetRange>,
+    ) -> object_store::Result<ObjectMeta> {
+        match entry.read_meta().await {
+            Ok(Some(meta)) => return Ok(meta),
```

Comment:
yeah, the current strategy is to cache the range at the first time of reading.

the current implementation of `maybe_prefetch_range` can be considered as a placeholder, it can be replaced with more sophisticated prefetch strategies at any time, and allow users to select a best prefetch strategy for them.